##  4.8: Grouping Data & Aggregating Variables

####  1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [1]:
#  Import the pandas, NumPy, and os libraries

import numpy as np
import pandas as pd
import os

In [2]:
# Paths to/from project folder

path_orig = r'C:\Users\kacor\CareerFoundry-Python\PROJECT - Instacart Basket Analysis\02 Data\Original Data'
path_prep = r'C:\Users\kacor\CareerFoundry-Python\PROJECT - Instacart Basket Analysis\02 Data\Prepared Data'

In [3]:
#  Import orders_products_merged (from 4-8Ex, from Prepared folder) 

df_ords_prods_merge = pd.read_pickle(os.path.join(path_prep, 'orders_products_merged_4-8Ex.pkl'))

In [4]:
df_ords_prods_merge.shape

(32404859, 17)

####  2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [5]:
# calculate the mean of the “order_number” column grouped by the “department_id” column

df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

####  3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

    The primary difference between the full dataframe and the subset is that there are department_ids present in the full dataset that are not present in the subset of data. I believe this is to be expected. Also, The averages of the subset data are different than thoses in the full dataset, as expected.

#### 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [6]:
#  First, create the max_order column

df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [7]:
df_ords_prods_merge.shape

(32404859, 17)

In [8]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_order_was_placed  \
0   2539329        1             1                   2                      8   
1   2398795        1             2                   3                      7   
2    473747        1             3                   3                     12   
3   2254736        1             4                   4                      7   
4    431534        1             5                   4                     15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  department_id  prices    price_range_loc   busiest_days  \
0         Soda              7     9.0  Mid-range product  Regularly day   
1         Soda              7     9.0  Mid-range product   Slowest days   
2         Soda              7     9.0  Mid-range product   Slowest days   
3         Soda              7     9.0  Mid-range product   Slowest days   
4         Soda              7     9.0  Mid-range product   Slowest days   

  busiest_period_of_day  max_order  loyalty_flag  
0        Average orders         10  New customer  
1        Average orders         10  New customer  
2           Most orders         10  New customer  
3        Average orders         10  New customer  
4           Most orders         10  New customer

In [9]:
#  Next, create the loyalty flag

df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
df_ords_prods_merge.shape

(32404859, 17)

In [11]:
df_ords_prods_merge.head(50)

order_id  user_id  order_number  orders_day_of_week  \
0    2539329        1             1                   2   
1    2398795        1             2                   3   
2     473747        1             3                   3   
3    2254736        1             4                   4   
4     431534        1             5                   4   
5    3367565        1             6                   2   
6     550135        1             7                   1   
7    3108588        1             8                   1   
8    2295261        1             9                   1   
9    2550362        1            10                   4   
10   2968173       15            15                   1   
11   1870022       15            17                   2   
12   1911383       15            18                   2   
13   2715276       15            21                   1   
14    487368       15            22                   1   
15   2293453       19             2                   5   
16   1973799       19             5                   6   
17    532817       19             7                   4   
18   1573906       21            10                   3   
19   1593000       31            10                   3   
20   2231262       31            17                   3   
21   2580647       43             6                   4   
22   2187180       43             9                   4   
23   2497897       52             1                   1   
24   1318871       52             2                   1   
25   1261384       52             3                   1   
26    580568       52             5                   2   
27   2428073       52             7                   2   
28    342306       52             9                   1   
29   1520353       52            10                   2   
30   2542086       52            11                   2   
31    944694       52            12                   1   
32   1268191       52            15                   2   
33   1122089       52            17                   2   
34   1498922       52            18                   2   
35    180919       52            19                   3   
36   2695875       52            21                   5   
37   3205960       67             1                   3   
38    970922       67             2                   1   
39    596585       67             3                   1   
40   2037087       67             4                   2   
41   2474111       67             5                   3   
42   1568481       67             6                   3   
43   2048527       67             8                   2   
44   2389031       67             9                   2   
45   2378562       67            10                   1   
46   1515698       67            11                   1   
47   1928050       67            12                   2   
48   1949274       67            13                   1   
49    551869       67            14                   1   

    hour_order_was_placed  days_since_prior_order  product_id  \
0                       8                     NaN         196   
1                       7                    15.0         196   
2                      12                    21.0         196   
3                       7                    29.0         196   
4                      15                    28.0         196   
5                       7                    19.0         196   
6                       9                    20.0         196   
7                      14                    14.0         196   
8                      16                     0.0         196   
9                       8                    30.0         196   
10                      9                     7.0         196   
11                     16                     8.0         196   
12                     11                     7.0         196   
13                      9                     7.0         196   
14                 

In [12]:
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

####  5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [13]:
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

OBSERVATION: New customers appear to by more expensive products and Loyal customers buy the least expensive products. The max price of 99999 for all three customer types is definately an error that must be investigated. 

In [14]:
# Investigate the prices column

df_ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

OBSERVATION: The median is 7.4 but the mean is 11.98 and the STD is 495 so it would appear that the max price is skewing the results.

In [15]:
# Check the values at the high end of the prices column

df_ords_prods_merge['prices'].drop_duplicates().sort_values(ascending = False)

29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

OBSERVATION: the prices of 99999 and 14900 are definately incorrect. There are no grocery items at the price level.

In [16]:
# Try to determine which items have the erroneous prices

df_ords_prods_merge.loc[df_ords_prods_merge['prices'].isin([99999, 14900])]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          hour_order_was_placed  days_since_prior_order  product_id  \
10030345                     14                     5.0       21553   
10030346                     16                     4.0       21553   
10030347                     21                    13.0       21553   
10030348                      8                    12.0       21553   
10030349                     19                    30.0       21553   
...                         ...                     ...         ...   
29166209                      8                     4.0       33664   
29166210                      9                     2.0       33664   
29166211                     15                     8.0       33664   
29166212                     20                     4.0       33664   
29166213                     11                     4.0       33664   

          add_to_cart_order  reordered                      product_name  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...                               ...   
29166209                  1          0             2 % Reduced Fat  Milk   
29166210                  1          1             2 % Reduced Fat  Milk   
29166211                 25          0             2 % Reduced Fat  Milk   
29166212                  8          0             2 % Reduced Fat  Milk   
29166213                 23          1             2 % Reduced Fat  Milk   

          department_id   prices     price_range_loc   busiest_days  \
10030345             16  14900.0  High-range product  Regularly day   
10030346             16  14900.0  High-range product  Regularly day   
10030347             16  14900.0  High-range product  Regularly day   
10030348             16  14900.0  High-range product   Busiest days   
10030349             16  14900.0  High-range product   Busiest days   
...                 ...      ...                 ...            ...   
29166209             16  99999.0  High-range product   Busiest days   
29166210             16  99999.0  High-range product   Busiest days   
29166211             16  99999.0  High-range product   Slowest days   
29166212             16  99999.0  High-range product   Slowest days   
29166213             16  99999.0  High-range product   Busiest days   

         busiest_period_of_day  max_order      loyalty_flag  
10030345           Most orders         40  Regular customer  
10030346           Most orders         40  Regular customer  
10030347        Average orders          4      New customer  
10030348        Average orders          4      New customer  
10030349        Average orders         16  Regular customer  
...                        ...        ...               ...  
29166209        Average orders         39  Regular customer  
29166210           Most orders         39  Regular customer  
29166211           Most orders         15  Regular c

In [17]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'].isin([99999, 14900])].describe()

order_id        user_id  order_number  orders_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000         5127.000000   
mean   1.708876e+06  106117.185684     15.302126            2.750536   
std    9.838956e+05   58594.982659     16.691849            2.010892   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.642320e+05   58136.000000      4.000000            1.000000   
50%    1.693799e+06  106021.000000      9.000000            3.000000   
75%    2.567244e+06  156389.000000     20.000000            5.000000   
max    3.421047e+06  206049.000000     99.000000            6.000000   

       hour_order_was_placed  days_since_prior_order    product_id  \
count            5127.000000             4776.000000   5127.000000   
mean               13.473376               11.899079  23201.815682   
std                 4.137326                9.068019   4153.742747   
min                 0.000000                0.000000  21553.000000   
25%                10.000000                5.000000  21553.000000   
50%                13.000000                8.000000  21553.000000   
75%                16.000000               16.000000  21553.000000   
max                23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered  department_id        prices  \
count        5127.000000  5127.000000         5127.0   5127.000000   
mean            7.986347     0.613614           16.0  26485.547494   
std             6.868889     0.486968            0.0  29186.636448   
min             1.000000     0.000000           16.0  14900.000000   
25%             3.000000     0.000000           16.0  14900.000000   
50%             6.000000     1.000000           16.0  14900.000000   
75%            11.000000     1.000000           16.0  14900.000000   
max            72.000000     1.000000           16.0  99999.000000   

         max_order  
count  5127.000000  
mean     29.791691  
std      24.929207  
min       3.000000  
25%      12.000000  
50%      21.000000  
75%      41.000000  
max      99.000000

OBSERVATION: There are well over 5000 items with erroneous prices. They appear to all be in department_id 16. 

In [18]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'].isin([14900])].describe()

order_id        user_id  order_number  orders_day_of_week  \
count  4.429000e+03    4429.000000   4429.000000         4429.000000   
mean   1.716950e+06  105002.918492     15.568526            2.727478   
std    9.830360e+05   59094.071587     17.160597            2.015750   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.678360e+05   56669.000000      4.000000            1.000000   
50%    1.714277e+06  103578.000000      9.000000            3.000000   
75%    2.572842e+06  156411.000000     20.000000            5.000000   
max    3.420827e+06  206049.000000     99.000000            6.000000   

       hour_order_was_placed  days_since_prior_order  product_id  \
count            4429.000000             4120.000000      4429.0   
mean               13.514789               11.963835     21553.0   
std                 4.152531                9.083407         0.0   
min                 0.000000                0.000000     21553.0   
25%                10.000000                5.000000     21553.0   
50%                14.000000                8.500000     21553.0   
75%                16.000000               17.000000     21553.0   
max                23.000000               30.000000     21553.0   

       add_to_cart_order    reordered  department_id   prices    max_order  
count        4429.000000  4429.000000         4429.0   4429.0  4429.000000  
mean            8.201174     0.614134           16.0  14900.0    30.916460  
std             6.922199     0.486854            0.0      0.0    25.632774  
min             1.000000     0.000000           16.0  14900.0     3.000000  
25%             3.000000     0.000000           16.0  14900.0    12.000000  
50%             6.000000     1.000000           16.0  14900.0    22.000000  
75%            11.000000     1.000000           16.0  14900.0    43.000000  
max            72.000000     1.000000           16.0  14900.0    99.000000

In [19]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'].isin([99999])].describe()

order_id        user_id  order_number  orders_day_of_week  \
count  6.980000e+02     698.000000    698.000000          698.000000   
mean   1.657648e+06  113187.514327     13.611748            2.896848   
std    9.885031e+05   54837.874488     13.224786            1.974941   
min    4.783000e+03     873.000000      1.000000            0.000000   
25%    8.215960e+05   71876.000000      4.000000            1.000000   
50%    1.549296e+06  114060.000000      9.000000            3.000000   
75%    2.511296e+06  155790.000000     19.000000            5.000000   
max    3.421047e+06  205227.000000     89.000000            6.000000   

       hour_order_was_placed  days_since_prior_order  product_id  \
count             698.000000              656.000000       698.0   
mean               13.210602               11.492378     33664.0   
std                 4.032515                8.966942         0.0   
min                 0.000000                0.000000     33664.0   
25%                10.000000                5.000000     33664.0   
50%                13.000000                8.000000     33664.0   
75%                16.000000               15.000000     33664.0   
max                23.000000               30.000000     33664.0   

       add_to_cart_order   reordered  department_id   prices   max_order  
count         698.000000  698.000000          698.0    698.0  698.000000  
mean            6.623209    0.610315           16.0  99999.0   22.654728  
std             6.358030    0.488028            0.0      0.0   18.366029  
min             1.000000    0.000000           16.0  99999.0    3.000000  
25%             2.000000    0.000000           16.0  99999.0   10.000000  
50%             5.000000    1.000000           16.0  99999.0   17.000000  
75%             9.000000    1.000000           16.0  99999.0   34.000000  
max            47.000000    1.000000           16.0  99999.0   99.000000

OBSERVATION: The problem appears to be with just two specific items in department_id 16:
- 4429 instances of product_id 21553 (Lowfat 2% Milkfat Cottage Cheese) with a price of 14900
- 698 instances of product_id 33664 (2 % Reduced Fat Milk) with a price of 99999

In [23]:
# See what prices look like in department_id 16 (dairy) without the abnormal prices. Try to determine what the prices of these 
# two items shoud be.

all_dairy = df_ords_prods_merge.loc[df_ords_prods_merge['department_id'].isin([16])]
all_dairy_clean = all_dairy.loc[~all_dairy['prices'].isin([99999, 14900])]
all_dairy_clean['prices'].describe()

count    5.393620e+06
mean     8.335056e+00
std      4.098207e+00
min      1.000000e+00
25%      4.800000e+00
50%      8.800000e+00
75%      1.190000e+01
max      1.500000e+01
Name: prices, dtype: float64

OBSERVATION: With erroneous prices removed, the max price in dairy is 15. This is a more realistic max price. 

In [31]:
# Determine prices for products similar to product_id 21153, containing the words (Lowfat 2% Milkfat Cottage Cheese) 

cottage_cheese_similar = all_dairy_clean.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") \
                          & product_name.str.contains("Milkfat")')
cottage_cheese_similar.sample(10)

order_id  user_id  order_number  orders_day_of_week  \
16190550   1904670    10536             3                   0   
16191200   1834753    73741            13                   3   
29898926    387376    44433            19                   5   
16192381   2709275   167147            41                   3   
30576119   1586081    45588            34                   1   
16192599    610025   186073             2                   3   
16191002   1040103    51232             1                   1   
16192157   2683174   141795             4                   3   
16191391   1404880    90466             2                   5   
16191186   2976052    72006            10                   4   

          hour_order_was_placed  days_since_prior_order  product_id  \
16190550                     16                     6.0       27790   
16191200                     10                    26.0       27790   
29898926                     16                     4.0        6852   
16192381                      4                     5.0       27790   
30576119                     18                     4.0       11203   
16192599                     15                    15.0       27790   
16191002                     22                     NaN       27790   
16192157                     15                    30.0       27790   
16191391                     14                     8.0       27790   
16191186                     10                     1.0       27790   

          add_to_cart_order  reordered  \
16190550                  5          0   
16191200                  1          1   
29898926                 18          1   
16192381                 21          0   
30576119                  5          1   
16192599                  2          0   
16191002                  2          0   
16192157                  8          0   
16191391                  3          1   
16191186                  5          0   

                                               product_name  department_id  \
16190550        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   
16191200        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   
29898926         4% Milkfat Cottage Cheese Small Curd 22 OZ             16   
16192381        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   
30576119  Cottage Doubles Peach, 2% Milkfat Lowfat Cotta...             16   
16192599        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   
16191002        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   
16192157        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   
16191391        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   
16191186        Small Curd Lowfat 2% Milkfat Cottage Cheese             16   

          prices    price_range_loc   busiest_days busiest_period_of_day  \
16190550     7.7  Mid-range product   Busiest days           Most orders   
16191200     7.7  Mid-range product   Slowest days           Most orders   
29898926     8.6  Mid-range product  Regularly day           Most orders   
16192381     7.7  Mid-range product   Slowest days         Fewest orders   
30576119     2.4  Low-range product   Busiest days        Average orders   
16192599     7.7  Mid-range product   Slowest days           Most orders   
16191002     7.7  Mid-range product   Busiest days        Average orders   
16192157     7.7  Mid-range product   Slowest days           Most orders   
16191391     7.7  Mid-range product  Regularly day           Most orders   
16191186     7.7  Mid-range product   Slowest days           Most orders   

          max_order      loyalty_flag  
16190550          9      New customer  
16191200         37  Regular customer  
29898926         66    Loyal customer  
16192381         44    Loyal customer  
30576119         34  Regular customer  
16192599         12  Regular customer  
16191002          9      New customer  
16192157          4      New cust

In [32]:
cottage_cheese_similar['prices'].describe()

count    3487.000000
mean        7.228219
std         1.757900
min         2.400000
25%         7.700000
50%         7.700000
75%         7.700000
max        11.800000
Name: prices, dtype: float64

OBSERVATION: The max price of items similar to cottage cheese is 11.80 and the mean price is 7.23. The mean price will be imputed for product_id 21553 (Lowfat 2% Milkfat Cottage Cheese) in place of 14900  

In [33]:
# Determine prices for products similar to product_id 33664, containing the words (2 % Reduced Fat Milk) 
reduced_fat_milk_similar = all_dairy_clean.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") \
                         & product_name.str.contains("Milk")')
reduced_fat_milk_similar.sample(10)

order_id  user_id  order_number  orders_day_of_week  \
20173100   2345955   202924            30                   4   
14537095    977316    17738            47                   4   
22332710   2949389   148296            21                   3   
25571053   1280294   163336            40                   2   
14551929   2672971   101547            50                   4   
14569277   2967739   195049            18                   6   
25570334    486304    29665            26                   4   
16034857   1399715   166390            14                   4   
16010445    740395    63789            92                   1   
8313220    1143782    71496             3                   4   

          hour_order_was_placed  days_since_prior_order  product_id  \
20173100                     10                     8.0       13166   
14537095                     10                     2.0       23909   
22332710                     14                     2.0       28593   
25571053                     13                    13.0       49319   
14551929                     18                     6.0       23909   
14569277                     15                     5.0       23909   
25570334                      9                     7.0       49319   
16034857                     14                     7.0        5785   
16010445                     17                     4.0        5785   
8313220                      10                    16.0       33572   

          add_to_cart_order  reordered  \
20173100                  2          1   
14537095                  5          1   
22332710                  2          1   
25571053                  6          0   
14551929                  4          1   
14569277                 13          0   
25570334                  3          1   
16034857                  2          1   
16010445                  4          1   
8313220                   1          1   

                                               product_name  department_id  \
20173100               Organic Milk Reduced Fat, 2% Milkfat             16   
14537095                                2% Reduced Fat Milk             16   
22332710                      Reduced Fat 2% Chocolate Milk             16   
25571053  Mozzarella String Cheese Made with 2% Reduced ...             16   
14551929                                2% Reduced Fat Milk             16   
14569277                                2% Reduced Fat Milk             16   
25570334  Mozzarella String Cheese Made with 2% Reduced ...             16   
16034857                        Organic Reduced Fat 2% Milk             16   
16010445                        Organic Reduced Fat 2% Milk             16   
8313220                    Lactose Free 2% Reduced Fat Milk             16   

          prices    price_range_loc   busiest_days busiest_period_of_day  \
20173100    12.8  Mid-range product   Slowest days           Most orders   
14537095     9.2  Mid-range product   Slowest days           Most orders   
22332710     1.7  Low-range product   Slowest days           Most orders   
25571053    11.8  Mid-range product  Regularly day           Most orders   
14551929     9.2  Mid-range product   Slowest days        Average orders   
14569277     9.2  Mid-range product  Regularly day           Most orders   
25570334    11.8  Mid-range product   Slowest days           Most orders   
16034857     7.9  Mid-range product   Slowest days           Most orders   
16010445     7.9  Mid-range product   Busiest days        Average orders   
8313220     10.5  Mid-range product   Slowest days           Most orders   

          max_order      loyalty_flag  
20173100         88    Loyal customer  
14537095         87    Loyal customer  
22332710         41    Loyal customer  
25571053         42    Loyal customer  
14551929         52    Loyal customer  
14569277         32  Regular customer  
25570334         44    Loyal customer  
16034857         20  Regular cust

In [34]:
reduced_fat_milk_similar['prices'].describe()

count    202238.000000
mean          9.086900
std           3.064869
min           1.200000
25%           7.900000
50%           9.200000
75%          11.400000
max          14.700000
Name: prices, dtype: float64

OBSERVATION: The mean is 9.09, the median 9.20 and the max is 14.70. With an erroneous price of 99999, I would suggest that 9.99 is a reasonable price for these items.  

In [35]:
# replace the price 14900 with 7.23, the mean price from cottage_cheese. Replace 99999 with 9.99  
df_ords_prods_merge['prices'] = df_ords_prods_merge['prices'].replace({99999: 9.99, 14900: 7.23})

count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

In [36]:
# Checks new results for prices column

df_ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

In [37]:
# Check spending habbits based on the 'loyalty_flag' with the corrected prices column

df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773527  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

OBSERVATION: The mean purchase price for all three customer types are nearly identical. The mean values are much lower and the max price is now 25

####  6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
    If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

    If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [38]:
#  Create column showing average spent for each user_id

df_ords_prods_merge['avg_spent'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [39]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_order_was_placed  \
0   2539329        1             1                   2                      8   
1   2398795        1             2                   3                      7   
2    473747        1             3                   3                     12   
3   2254736        1             4                   4                      7   
4    431534        1             5                   4                     15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  department_id  prices    price_range_loc   busiest_days  \
0         Soda              7     9.0  Mid-range product  Regularly day   
1         Soda              7     9.0  Mid-range product   Slowest days   
2         Soda              7     9.0  Mid-range product   Slowest days   
3         Soda              7     9.0  Mid-range product   Slowest days   
4         Soda              7     9.0  Mid-range product   Slowest days   

  busiest_period_of_day  max_order  loyalty_flag  avg_spent  
0        Average orders         10  New customer   6.367797  
1        Average orders         10  New customer   6.367797  
2           Most orders         10  New customer   6.367797  
3        Average orders         10  New customer   6.367797  
4           Most orders         10  New customer   6.367797

In [40]:
#  Create spending flag column

df_ords_prods_merge.loc[df_ords_prods_merge['avg_spent'] < 10, 'spending_flag'] = 'Low spender'
df_ords_prods_merge.loc[df_ords_prods_merge['avg_spent'] >= 10, 'spending_flag'] = 'High spender'

In [41]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_order_was_placed  \
0   2539329        1             1                   2                      8   
1   2398795        1             2                   3                      7   
2    473747        1             3                   3                     12   
3   2254736        1             4                   4                      7   
4    431534        1             5                   4                     15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  department_id  prices    price_range_loc   busiest_days  \
0         Soda              7     9.0  Mid-range product  Regularly day   
1         Soda              7     9.0  Mid-range product   Slowest days   
2         Soda              7     9.0  Mid-range product   Slowest days   
3         Soda              7     9.0  Mid-range product   Slowest days   
4         Soda              7     9.0  Mid-range product   Slowest days   

  busiest_period_of_day  max_order  loyalty_flag  avg_spent spending_flag  
0        Average orders         10  New customer   6.367797   Low spender  
1        Average orders         10  New customer   6.367797   Low spender  
2           Most orders         10  New customer   6.367797   Low spender  
3        Average orders         10  New customer   6.367797   Low spender  
4           Most orders         10  New customer   6.367797   Low spender

In [42]:
#  Check distributin of spending flag column

df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     32285150
High spender      119709
Name: spending_flag, dtype: int64

####  7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”

If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”

If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [43]:
# create order_freq column based on the median of days_since_prior_order for each user_id

df_ords_prods_merge['order_freq'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [44]:
# Verify output of order frequency column

df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_order_was_placed  \
0   2539329        1             1                   2                      8   
1   2398795        1             2                   3                      7   
2    473747        1             3                   3                     12   
3   2254736        1             4                   4                      7   
4    431534        1             5                   4                     15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  department_id  prices    price_range_loc   busiest_days  \
0         Soda              7     9.0  Mid-range product  Regularly day   
1         Soda              7     9.0  Mid-range product   Slowest days   
2         Soda              7     9.0  Mid-range product   Slowest days   
3         Soda              7     9.0  Mid-range product   Slowest days   
4         Soda              7     9.0  Mid-range product   Slowest days   

  busiest_period_of_day  max_order  loyalty_flag  avg_spent spending_flag  \
0        Average orders         10  New customer   6.367797   Low spender   
1        Average orders         10  New customer   6.367797   Low spender   
2           Most orders         10  New customer   6.367797   Low spender   
3        Average orders         10  New customer   6.367797   Low spender   
4           Most orders         10  New customer   6.367797   Low spender   

   order_freq  
0        20.5  
1        20.5  
2        20.5  
3        20.5  
4        20.5

In [45]:
# Create order frequency flag column 

df_ords_prods_merge.loc[df_ords_prods_merge['order_freq'] > 20, 'order_freq_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['order_freq'] > 10) & (df_ords_prods_merge['order_freq'] <= 20), 'order_freq_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['order_freq'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [46]:
# Check the breakdown of new flag column

df_ords_prods_merge['order_freq_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_freq_flag, dtype: int64

In [47]:
df_ords_prods_merge.shape

(32404859, 21)

In [48]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 21 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   hour_order_was_placed   int64  
 5   days_since_prior_order  float64
 6   product_id              int64  
 7   add_to_cart_order       int64  
 8   reordered               int64  
 9   product_name            object 
 10  department_id           int64  
 11  prices                  float64
 12  price_range_loc         object 
 13  busiest_days            object 
 14  busiest_period_of_day   object 
 15  max_order               int64  
 16  loyalty_flag            object 
 17  avg_spent               float64
 18  spending_flag           object 
 19  order_freq              float64
 20  order_freq_flag         object 
dtypes: float64(4), int64(10), obj

In [49]:
#  Export df_ords_prods_merge to Prepared folder in .pkl format  

df_ords_prods_merge.to_pickle(os.path.join(path_prep, 'orders_products_merged_4-8Tsk.pkl'))